src/halmos/mapper.py

```diff
@@ -78,42 +77,28 @@ def get_contract_mapping_info_by_bytecode(
         return None
 
     def append_node(self, contract_name: str, node: AstNode):
-        contract_mapping_info = self.get_contract_mapping_info_by_name(contract_name)
-
-        if contract_mapping_info is None:
-            raise ValueError(f"Contract {contract_name} not found")
-
+        contract_mapping_info = self._contracts[contract_name]
         contract_mapping_info.nodes.append(node)
 
-    def parse_ast(self, node: Dict, contract_name: str = ""):
+    def parse_ast(self, node: Dict, contract_name: str):
         node_type = node["nodeType"]
-
         if node_type in self._PARSING_IGNORED_NODE_TYPES:
             return
 
-        current_contract = self._get_current_contract(node, contract_name)
-
         if node_type == "ContractDefinition":
-            if current_contract not in self._contracts:
-                self.add_contract_mapping_info(
-                    contract_name=current_contract, bytecode="", nodes=[]
-                )
-
-            if self.get_contract_mapping_info_by_name(current_contract).nodes:
+            if self.get_or_create(contract_name).nodes:
                 return
+
         elif node_type != "SourceUnit":
             id, name, address, visibility = self._get_node_info(node, node_type)
-
-            self.append_node(
-                current_contract,
-                AstNode(node_type, id, name, address, visibility),
-            )
+            ast_node = AstNode(node_type, id, name, address, visibility)
+            self.append_node(contract_name, ast_node)
 
         for child_node in node.get("nodes", []):
-            self.parse_ast(child_node, current_contract)
+            self.parse_ast(child_node, contract_name)
```

Comment:
this doesn't seem semantic-preserving. why not treating differently the case of `child_node["nodeType"] == "ContractDefinition"`? in that case, `child_node.get("name", "")` was used instead of `contract_name`.

Reply:
it is definitely not semantic preserving, because I think there was an issue. The `except Exception: pass` was swallowing exceptions like this:

```
error parsing symbols for contract FuzzCantCatch
Traceback (most recent call last):
  File "/Users/karma/projects/halmos/src/halmos/__main__.py", line 1304, in parse_symbols
    Mapper().parse_ast(contract_map[contract_name][0]["ast"])
  File "/Users/karma/projects/halmos/src/halmos/mapper.py", line 103, in parse_ast
    self.parse_ast(child_node, current_contract)
  File "/Users/karma/projects/halmos/src/halmos/mapper.py", line 97, in parse_ast
    self.append_node(
  File "/Users/karma/projects/halmos/src/halmos/mapper.py", line 79, in append_node
    raise ValueError(f"Contract {contract_name} not found")
ValueError: Contract  not found
```

Reply:
Now you're right, I had some wrong assumptions about what we would find in the AST. For instance with this contract: https://github.com/karmacoma-eth/halmos-sandbox/blob/main/test/46_hellFunc.t.sol (one solidity file with no eponymous contract, and it contains 2 contract definitions)

Walking the AST in `out/46_hellFunc.t.sol/FuzzCantCatch.json` we find:

```
SourceUnit
  PragmaDirective
  UserDefinedValueTypeDefinition (Int)
  UsingForDirective
  UsingForDirective
  UsingForDirective
  UsingForDirective
  FunctionDefinition (add)
    Block
  FunctionDefinition (div)
    Block
  FunctionDefinition (mul)
    Block
  FunctionDefinition (sub)
    Block
  ContractDefinition (FuzzCantCatch)
    VariableDeclaration (numbr)
    VariableDeclaration (namber)
    VariableDeclaration (nunber)
    VariableDeclaration (mumber)
    VariableDeclaration (numbor)
    VariableDeclaration (numbir)
    FunctionDefinition (hellFunc)
      Block
    FunctionDefinition (_numbaar)
      Block
  ContractDefinition (HellCatcher)
    VariableDeclaration (fuzz)
    FunctionDefinition (setUp)
      Block
    FunctionDefinition (test_hellFunc)
      Block
    FunctionDefinition (testCatchEasyBugWithFuzz)
      Block
```

And if we walk the AST of `out/46_hellFunc.t.sol/HellCatcher.json`, we get exactly the same thing 😵 it's actually the AST of the whole source file, not just the contract

Reply:
Still, based on this I think we should indeed "guess" `contract_name` based on the json filename, and dig into the AST until we hit the right node. 

We still have a couple problems:
- how to handle another contract definition that's not the same as `contract_name`? Currently it would be processed if it has not been seen yet, or skipped if it has
- how to handle contract name collisions (A.sol and B.sol both define a contract called C), I think right now it's nondeterministic (whichever one is processed first is added, the other one is skipped)

Reply:
ok I found what was causing the initial issue, in `46_hellFunc.t.sol`:

```solidity
type Int is uint256;
```

What happens is this:
- it is outside of a contract definition
- it results in an AST node of type `UserDefinedValueTypeDefinition`
- because `node_type != "SourceUnit"`, we try to append it to the `current_contract` mapping
- but because we haven't encountered a `ContractDefinition` yet, it tries to append it to the default `contract_name = ""`, which raises the exception

Similar thing going on here: https://github.com/karmacoma-eth/halmos-sandbox/blob/main/test/17_pAtHExPlOSion.t.sol, the free functions (`FunctionDefinition` AST nodes) before the contract definition cause the same exception